Update dashboard and card api POST and PUT endpoints to return as GET

markbastian opened the pull request:
In both the `metabase.api.dashboard` and `metabase.api.card` nses the primary change was to add a `hydrate-dashboard|card-details` function that is shared across the GET, POST, and PUT operations so that all 3 return an object of the same shape, conforming to the GET response. This removes the need for a POST|PUT + GET call on the FE. Note the in all cases, the returned value was nearly the same as the GET response, so it really was returning _almost_ the same thing so this does effectively halve the load required for these operations.

Unit tests were added to check that the returned objects from POST|PUT are the same as from a GET.

markbastian committed:
Adding stub for POST+GET same result

This commit just stubs in the `post-same-as-get-test` test in `metabase.api.dashboard-test`.

markbastian committed:
Hydrating POST /api/dashboard results to = GET

This is the first stage of "PUT and POST and endpoint should return data just like calling GET #34828" in which a POST to create a dashboard also hydrates the details needed to return the identical results as if it were a GET.

This enables a single API call vs. two (a POST and and immediate GET).

markbastian committed:
Adding updated value to PUT /api/dashboard response

This change adds the same hydration logic to PUT responses as POST responses.

Also added a span around `update-dashboard` to have parity with `get-dashboard`.

markbastian committed:
Changed span name for consistency

markbastian committed:
Updating create-dashboard-test to include GET comparison

To be consistent with the PUT test this just combines two tests for dashboard creation into one.

markbastian committed:
POST/PUT api/card now returns same value as GET

The primary change is using the same hydration logic across the board.

This ensures the same object shape is returned for all operations an removes the need for a GET immediately after a POST or PUT when a create or update action occurs.

markbastian committed:
Removing GET on Dashboard Save

When a dashboard is updated via PUT, the succeeding GET operation is no longer needed and has been removed.

markbastian committed:
Restored fetch unrelated to dashboard creation.

markbastian committed:
Adding =? to `create-a-card` test.

adam-james-v commented on test/metabase/api/card_test.clj:
```diff
@@ -674,19 +674,26 @@
+                                                dataset-query))
+                retrieved     (mt/user-http-request :rasta :get 200 (str "card/" card-id))]
+            (is (pos-int? card-id))
+            (testing "A POST returns the newly created object, so no follow-on GET is required"
```
This PR addresses a bug right? I've seen #ISSUENUM put in testing strings as a reference before, if you want to add that here.

markbastian replied:
Hmm, more of a performance improvement, but i like the idea. Will update.

adam-james-v reviewed (approved):
Backend changes look nice. I like that hydration is pulled into one fn. Makes it easy to reuse in all the right spots.

markbastian committed:
Updated testing forms to reflect issue being addressed (#34828).

markbastian committed:
Reduced expected e2e load count for 21556 test

This e2e test was example based and expected a PUT + 3 additional GETs. With the new code you only need to do 2 GETs since the PUT result is used.

calherries commented on src/metabase/api/card.clj:
```diff
@@ -166,15 +185,7 @@
-                             [:moderation_reviews :moderator_details])
-                 collection.root/hydrate-root-collection
+                 hydrate-card-details
                  (cond-> ;; card
                    (:dataset raw-card) (t2/hydrate :persisted))
```
Just spotted you don't need this anymore. You're doing it in `hydrate-card-details`.
```suggestion
```

calherries commented on src/metabase/api/card.clj:
```diff
@@ -166,15 +185,7 @@
+                 hydrate-card-details
                  (cond-> ;; card
                    (:dataset raw-card) (t2/hydrate :persisted))
                  (last-edit/with-last-edit-info :card))]
```
Why isn't `(last-edit/with-last-edit-info :card)` part of `hydrate-card-details`?

markbastian replied:
I was wondering about this as well. It's the same situation in api.dashboard in terms of when to update the last edit info. It does collect the logic in one place, but it feels like it isn't part of the hydration logic. WDYT?

calherries reviewed (changes requested):
Nice! I was about to ping you to try something like this because I just merged [Move functions for creating and updating cards from api to models namespace](https://github.com/metabase/metabase/pull/36123). Unfortunately it creates conflicts with this in `metabase.api.card`. In that PR I consolidated a lot of the stuff in your `hydrate-card-details` in a similar function called `hydrate-for-frontend`, but deliberately left out some keys because I didn't want to make too many changes outside the refactoring I was doing. It should be pretty straightfoward to resovle the conflict though

calherries commented on src/metabase/api/card.clj:
```diff
@@ -129,6 +130,24 @@
+                    [:collection :is_personal]
+                    [:moderation_reviews :moderator_details])
+        (cond-> dataset (t2/hydrate :persisted))
+        collection.root/hydrate-root-collection)))
```
Are you sure the root collection needs to be hydrated on all endpoints? I would expect that all collections are normalized on the frontend and cached separately from the card stuff, including the root collection.

markbastian replied:
Agreed. Done.

markbastian replied:
Per our discussion, let's go with removing. This means we'll not have _identical_ results as POST|PUT + GET, but the assumption is that the `:collection` will be present already on the FE in those cases.

markbastian committed:
Renaming fns and reducing root collection hydration for cards

Rebased on master and merged with @calherries `metabase.api.card` changes.

Renamed `hydrate-for-frontend` to `hydrate-card-details` and added a span for observability.

Removed `collection.root/hydrate-root-collection` which makes cards returned from POST and PUT not identical to GET, but only because the `:collection` key is missing, which should already be present on the FE in these situations. Tests were updated to reflect this.

markbastian committed:
Reverting FE changes

The FE changes for this are not a subtle as simply removing a follow on GET. It makes more sense to let the BE changes get merged then the the FE team work on leveraging those changes.

markbastian commented:
Fixes https://github.com/metabase/metabase/issues/34828. Note that FE portions of that ticket were moved to https://github.com/metabase/metabase/issues/36326.
